### PR TITLE
Add warning for query passthrough

### PR DESCRIPTION
**Pull request by Jessie212**
## Description

Create fragment and warning callout for query passthrough workaround

## Release notes

(s) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: docs/src/main/sphinx/connector/postgresql.rst [hunk 1]
```diff
@@ -342,7 +342,14 @@ The ``query`` function allows you to query the underlying database directly. It
 requires syntax native to PostgreSQL, because the full query is pushed down and
 processed in PostgreSQL. This can be useful for accessing native features which
 are not available in Trino or for improving query performance in situations
-where running a query natively may be faster.
+where running a query natively may be fasters.
+
+.. include:: query-passthrough-warning.fragment
+
+.. Warning::
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/postgresql.rst:349]: Move this all into the fragment

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment [hunk 2]
```diff
@@ -0,0 +1,6 @@
+The native query passed to the underlying data source must return a table as a
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/query-passthrough-warning.fragment:1]: Reword more towards the draft I gave you to avoid SQL syntax.

**Review by mosabua — CHANGES_REQUESTED** [review]
Add the following:

* Include in all relevant connectors
* Keep it all in the fragment
* Change commit message and PR message to be about the content .. not the mechanics of how you do it (the fragment approach)

File: docs/src/main/sphinx/connector/postgresql.rst [hunk 3]
```diff
@@ -342,7 +342,14 @@ The ``query`` function allows you to query the underlying database directly. It
...
+
+  Using this method allows any query that reads data to bypass access control,
+  as no validation or security checks are performed.
```
Comment (hashhar) [inline docs/src/main/sphinx/connector/postgresql.rst:352]: It might be better to say that only security checks defined in the source system apply?

i.e. permissions from Trino won't apply but any permissions you have configured on the source system still apply.

Comment (mosabua) [reply docs/src/main/sphinx/connector/postgresql.rst:352]: Yeah ... the draft I had for @Jessie212 had that. Still working out the final wording.

File: docs/src/main/sphinx/connector/postgresql.rst (hunk 1 above)
Comment (Jessie212) [reply docs/src/main/sphinx/connector/postgresql.rst:349]: I'm trying to find a way. The normal `.. Warning::` syntax won't work inside a .fragment file.

File: docs/src/main/sphinx/connector/postgresql.rst (hunk 3 above)
Comment (Jessie212) [reply docs/src/main/sphinx/connector/postgresql.rst:352]: @mosabua I'm just going to use your original draft, but with slight changes to the last paragraph.

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment [hunk 4]
```diff
@@ -0,0 +1,11 @@
+
+.. Warning::
+
+    The native query passed to the underlying data source is required to return
+    a table as a result set. This can be a one column table with one row. No
+    validation or security checks of these queries are performed. Therefore, any
+    queries reading data can be used. Queries that ``MODIFY``, ``UPDATE``,
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/query-passthrough-warning.fragment:7]: Do not use code for modify, update, insert, or delete

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+    ``INSERT``, or ``DELETE`` data, however, cannot be used since they do not
+    return a table. However, if supported by the underlying data
+    source/database, you can enable such queries by appending a statement with a
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/query-passthrough-warning.fragment:10]: just use data source only

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment [hunk 6]
```diff
@@ -0,0 +1,11 @@
+
+.. Warning::
+
+    The native query passed to the underlying data source is required to return
+    a table as a result set. This can be a one column table with one row. No
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/query-passthrough-warning.fragment:5]: Lets move the sentence to the end of the paragraph and modify to  "No validation or security checks of these queries are performed by Trino. Only the data source performs these task using its own configuration."

Comment (Jessie212) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment:5]: I will take on these notes, but should I also format the section as it was previously, with a rest of the section just as paragraph and these last two sentences as the warning? It seems to me that the rest is just explaining the workaround process and should not be included in the warning.

**Comment by mosabua** [comment]
Drop the "workaround" from the PR title and commit message please

File: docs/src/main/sphinx/connector/postgresql.rst [hunk 7]
```diff
@@ -342,7 +342,9 @@ The ``query`` function allows you to query the underlying database directly. It
 requires syntax native to PostgreSQL, because the full query is pushed down and
 processed in PostgreSQL. This can be useful for accessing native features which
 are not available in Trino or for improving query performance in situations
-where running a query natively may be faster.
+where running a query natively may be fasters.
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/postgresql.rst:345]:
```suggestion
where running a query natively may be faster.
```

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment [hunk 8]
```diff
@@ -0,0 +1,8 @@
+The native query passed to the underlying data source is required to return a
+table as a result set. This can be a one column table with one row. No
+validation or security checks of these queries are performed by Trino. Only the
+data source performs these task using its own configuration. You can use that
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/query-passthrough-warning.fragment:4]:
```suggestion
data source performs these tasks using its own configuration. You can use passthrough queries to
```

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+read data, however queries that modify, update, insert, or delete data, cannot
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/query-passthrough-warning.fragment:5]:
```suggestion
read data, however queries that update, insert, or delete data cannot
```

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment [hunk 10]
```diff
@@ -0,0 +1,8 @@
...
+be used since they do not return a table. If supported by the underlying data
+source, you can enable such queries by appending a statement with a function
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/query-passthrough-warning.fragment:7]:
```suggestion
source, you can enable such queries by appending a statement
```

**Review by mosabua — CHANGES_REQUESTED** [review]

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment [hunk 11]
```diff
@@ -0,0 +1,8 @@
+The native query passed to the underlying data source is required to return a
+table as a result set. This can be a one column table with one row. No
+validation or security checks of these queries are performed by Trino. Only the
+data source performs these tasks using its own configuration. You can use
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/query-passthrough-warning.fragment:4]: Maybe break up paragraph here

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+passthrough queries to read data; however, queries that update, insert, or
+delete data cannot be used since they do not return a table. If supported by the
+underlying data source, you can enable such queries by appending a statement
+with ``RETURNING 1``.
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/query-passthrough-warning.fragment:8]: Maybe we should add afterwards 

These characteristics can result in undesired data access or data loss.

Or so

Comment (Jessie212) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment:8]: Yes. That's the kind of warning I wanted to add. Should I create a callout box?

Comment (mosabua) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment]: No .. just have separate paragraphs and then lets see what reviewers think. Personally I think its fine to just use normal paragraphs.

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment [hunk 13]
```diff
@@ -0,0 +1,10 @@
+The native query passed to the underlying data source is required to return a
+table as a result set. This can be a one column table with one row. No
+validation or security checks of these queries are performed by Trino. The
+data source performs these tasks using its own configuration.
+
+You can use passthrough queries to read data; however, queries that update,
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/query-passthrough-warning.fragment:6]: lets change this to 

queries that write data typically can not be used since they do not return a table.

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment [hunk 14]
```diff
@@ -0,0 +1,10 @@
+The native query passed to the underlying data source is required to return a
+table as a result set. This can be a one column table with one row. No
+validation or security checks of these queries are performed by Trino. The
+data source performs these tasks using its own configuration.
+
+You can use passthrough queries to read data; however, queries that write data
```
Comment (electrum) [inline docs/src/main/sphinx/connector/query-passthrough-warning.fragment:6]: This implies that writes are not possible unless the underlying database supports `RETURNING`. This seems like too strong a statement to make. It seems more likely that in most databases, it would be possible to execute the write on the underlying database, even if the Trino query ends up failing due to no result set.

Comment (jhlodin) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment:6]: So instead of saying we don't support writes, we need to better describe that it may appear to fail but still be processed since that's a worse outcome than just failing outright.

Suggest: "You can use passthrough queries to read data; however, queries that write data typically cannot be used since they do not return a table. If supported by the underlying data source, you can enable such queries by..."
->
"You can use passthrough queries to read data; however, queries that write data typically do not return a table, so when a write operation is passed Trino declares that the query failed even if the database performs the operation. If supported by the underlying data source, you can avoid this outcome by..."

**Review by jhlodin — APPROVED** [review]
LGTM

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment [hunk 15]
```diff
@@ -0,0 +1,11 @@
+The native query passed to the underlying data source is required to return a
+table as a result set. This can be a one column table with one row. No
```
Comment (hashhar) [inline docs/src/main/sphinx/connector/query-passthrough-warning.fragment:2]:
```suggestion
table as a result set. This can be a one column table with one row. No
```

Why is this sentence required? It makes it sound like other "shapes" aren't ideal.

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+validation or security checks of these queries are performed by Trino. The
+data source performs these tasks using its own configuration.
+
+You can use passthrough queries to read data; however, queries that write data
+typically do not return a table. As a result, when a write operation is passed,
+Trino declares that the query failed even if the database performs the
+operation. If supported by the underlying data source, you can avoid this
+outcome by appending a statement with ``RETURNING 1``. Performing this action
+may grant access to restricted data or result in data loss.
```
Comment (hashhar) [inline docs/src/main/sphinx/connector/query-passthrough-warning.fragment:11]: is this factual? IIRC the query fails during ANALYSIS itself so there's no possibility of a DELETE/UPDATE getting executed if the Trino query shows a failure. @kasiafi can you please double check my understanding?

**Review by hashhar — CHANGES_REQUESTED** [review]
LGTM % fact-check

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment (hunk 16 above)
Comment (kasiafi) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment:11]: @hashhar I don't think there's a possibility that an operation (like DELETE or UPDATE) got executed though Trino query failed. I would consider that a serious issue.

All operations that do not return a result set fail during analysis, so they never reach execution.

Please note that the success or failure of a query pass-through depends directly on the *presence of a result set*, not on the *operation type*. Theoretically, if some DB returned a result set metadata for a DELETE, we would be happy to execute it. cc @mosabua @Jessie212

Comment (mosabua) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment:11]: Thats exactly the point of the warning ... however ... do we know that the query was actually not run on the underlying system .. @electrum was mentioning that maybe a query like a DELETE query is actually passed through .. and run .. and then just fails on the trino side since no result set is returned.. but the query actually deleted records.

So the aim of this PR is to add a warning of whatever the potential behavior is. Could you verify ?

Comment (hashhar) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment:11]: No, the failure happens during analysis of the query. There's no possibility of running a query if it doesn't return a return set. We check whether result set would be returned BEFORE executing the query.

You can verify this behaviour through the various QueryRunners we have (PostgreSqlQueryRunner/MySqlQueryRunner were the ones I verified).

Comment (kasiafi) [inline docs/src/main/sphinx/connector/query-passthrough-warning.fragment:11]: I would remove "may grant access to restricted data". If you mean access which would be restricted if the query was run in Trino, then it is equally true for read and write operations.

Regarding "result in data loss", I would reword it. I a user runs a DELETE query, it is their expectation that data will be deleted. "Data loss" sounds like a bug.

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment [hunk 17]
```diff
@@ -0,0 +1,11 @@
+The native query passed to the underlying data source is required to return a
+table as a result set. This can be a one column table with one row. No
+validation or security checks of these queries are performed by Trino. The
+data source performs these tasks using its own configuration.
+
+You can use passthrough queries to read data; however, queries that write data
+typically do not return a table. As a result, when a write operation is passed,
+Trino declares that the query failed even if the database performs the
+operation. If supported by the underlying data source, you can avoid this
+outcome by appending a statement with ``RETURNING 1``. Performing this action
```
[the remark this answers is not in the document]
Comment (hashhar) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment:10]: Postgres has `RETURNING`, SQL Server has `OUTPUT`. I don't know about others.

[the remark this answers is not in the document]
Comment (hashhar) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment:10]: but yes, this statement won't work as it is for all connectors.

[the remark this answers is not in the document]
Comment (jhlodin) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment:10]: What connectors support what syntax? Can we get a list?

[the remark this answers is not in the document]
Comment (mosabua) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment:10]: Thats basically the documentation for the underlying database .. its a pass through .. so we cant make a list .. we can only refer to the docs of the data source.

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment (hunk 16 above)
Comment (mosabua) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment:11]: How is that even possible .. Trino verifies a random pass through query string that is valid for the underlying data source without running the query? Does it pass it through for analysis only or something?

Comment (mosabua) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment:11]: How is that even possible .. Trino verifies a random pass through query string that is valid for the underlying data source without running the query? Does it pass it through for analysis only or something? Or does Trino only validate the larger query .. but not the pass through bit .. and if thats the case .. how does it know if that returns a result set or whatever else.

Comment (martint) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment:11]: We should discourage using these table functions for executing queries that might have side-effects in the underlying database. Doing that can be problematic in the face of query or task retries if the operation is not idempotent.

Comment (mosabua) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment:11]: How do we word that so users understand? You are just talking about write access and DDL stuff right? Or any other side effects?

Should we just say that users should use this for just for read-only access?

File: docs/src/main/sphinx/connector/query-passthrough-warning.fragment (hunk 15 above)
Comment (mosabua) [reply docs/src/main/sphinx/connector/query-passthrough-warning.fragment]: @Jessie212 will remove

**Review by mosabua — APPROVED** [review]
I think this is good to go now. Please look and merge @kasiafi or @hashhar

**Review by kasiafi — APPROVED** [review]

**Review by hashhar — APPROVED** [review]
